tools/outline_proxy_controller/network_monitor.cpp

```diff
@@ -0,0 +1,148 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <array>
+#include <cstddef>      // size_t, byte
+#include <cstring>      // memset
+#include <system_error>
+#include <type_traits>  // underlying_type_t, is_integral
+
+#include <linux/netlink.h>
+#include <linux/rtnetlink.h>
+
+#include <boost/asio.hpp>
+
+#include "logger.h"
+#include "network_monitor.h"
+
+using namespace outline;
+
+static const std::size_t kRecvBufferSize = 4096;
+
+NetworkMonitor::NetworkMonitor(const boost::asio::any_io_executor &io_context)
+  : netlink_socket_{io_context}
+{
+  using namespace boost::asio::generic;
+
+  sockaddr_nl sa;
+  std::memset(&sa, 0, sizeof(sa));
+  sa.nl_family = AF_NETLINK;
+  sa.nl_groups = RTMGRP_LINK         // network card change
+               | RTMGRP_IPV4_IFADDR  // IPv4 address change
+               | RTMGRP_IPV4_ROUTE   // IPv4 route table change
+               | RTMGRP_IPV6_IFADDR  // IPv6 address change
+               | RTMGRP_IPV6_ROUTE;  // IPv6 route table change
+
+  netlink_socket_.open({AF_NETLINK, NETLINK_ROUTE});
+  netlink_socket_.bind({&sa, sizeof(sa)});
+
+  // for non-blocking peek IsNextChangeEventReady
+  netlink_socket_.non_blocking(true);
+
+  logger.info("network monitor initialized");
+}
+
+NetworkMonitor::~NetworkMonitor() {
+  logger.info("network monitor destroyed");
+}
+
+boost::asio::awaitable<NetworkChangeEvent> NetworkMonitor::WaitForChangeEvent() {
+  using namespace boost::asio;
+
+  std::array<std::byte, kRecvBufferSize> buf;
+  NetworkChangeEvent received_events;
+
+  do {
+    auto len = co_await netlink_socket_.async_receive(buffer(buf), use_awaitable);
+
+    // https://linux.die.net/man/7/netlink
+    // https://man7.org/linux/man-pages/man7/rtnetlink.7.html
+    for (auto msg = reinterpret_cast<nlmsghdr*>(buf.data()); NLMSG_OK(msg, len); msg = NLMSG_NEXT(msg, len)) {
+      switch (msg->nlmsg_type) {
+        case NLMSG_DONE:
+          goto EndOfMultiPartMsg;
+        case NLMSG_ERROR:
+        {
+          auto err = reinterpret_cast<nlmsgerr*>(NLMSG_DATA(msg));
+          throw std::system_error{err->error, std::system_category()};
+        }
+        case RTM_NEWLINK:
+          received_events.SetNicChanged();
+          break;
+        case RTM_NEWADDR:
+        case RTM_DELADDR:
+          received_events.SetAddressChanged();
+          break;
+        case RTM_NEWROUTE:
+        case RTM_DELROUTE:
+          received_events.SetRoutingChanged();
+          break;
+      }
+    }
+
+EndOfMultiPartMsg:
+    ;
+  } while (received_events.IsEmpty() || IsNextChangeEventReady());
+
+  co_return received_events;
+}
+
+bool NetworkMonitor::IsNextChangeEventReady() {
+  std::array<std::byte, 1> buf;
+  boost::system::error_code err;
+
+  // receive won't block because we already set `netlink_socket_.non_blocking(true)`
+  auto len = netlink_socket_.receive(boost::asio::buffer(buf), netlink_socket_.message_peek, err);
```

Comment:
Won't this remove the event from the buffer? Will the `netlink_socket_.async_receive()` call in `NetworkMonitor::WaitForChangeEvent` still see the event?

Reply:
Yes it will still stay in the buffer, because I used [`message_peek` flag](https://www.boost.org/doc/libs/1_80_0/doc/html/boost_asio/reference/basic_stream_socket/message_peek.html) here:

> Peek at incoming data without removing it from the input queue.